sql-plugin/src/main/java/com/nvidia/spark/rapids/iceberg/parquet/ApplyNameMapping.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.OriginalType;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Type;
+import org.apache.parquet.schema.Types;
+
+/** Derived from Apache Iceberg's ApplyNameMapping Parquet support class. */
+public class ApplyNameMapping extends ParquetTypeVisitor<Type> {
+  private static final String LIST_ELEMENT_NAME = "element";
+  private static final String MAP_KEY_NAME = "key";
+  private static final String MAP_VALUE_NAME = "value";
+  private final NameMapping nameMapping;
+  private final Deque<String> fieldNames = Lists.newLinkedList();
+
+  ApplyNameMapping(NameMapping nameMapping) {
+    this.nameMapping = nameMapping;
+  }
+
+  @Override
+  public Type message(MessageType message, List<Type> fields) {
+    Types.MessageTypeBuilder builder = Types.buildMessage();
+    fields.stream().filter(Objects::nonNull).forEach(builder::addField);
+
+    return builder.named(message.getName());
+  }
+
+  @Override
+  public Type struct(GroupType struct, List<Type> types) {
+    MappedField field = nameMapping.find(currentPath());
+    List<Type> actualTypes = types.stream().filter(Objects::nonNull).collect(Collectors.toList());
+    Type structType = struct.withNewFields(actualTypes);
+
+    return field == null ? structType : structType.withId(field.id());
+  }
+
+  @Override
+  public Type list(GroupType list, Type elementType) {
+    Preconditions.checkArgument(elementType != null,
+        "List type must have element field");
+
+    Type listElement = ParquetSchemaUtil.determineListElementType(list);
+    MappedField field = nameMapping.find(currentPath());
+
+    Types.GroupBuilder<GroupType> listBuilder = Types.buildGroup(list.getRepetition())
+        // Spark 3.1 uses Parquet 1.10 which does not have LogicalTypeAnnotation
+//        .as(LogicalTypeAnnotation.listType());
+        .as(OriginalType.LIST);
+    if (listElement.isRepetition(Type.Repetition.REPEATED)) {
+      listBuilder.addFields(elementType);
+    } else {
+      listBuilder.repeatedGroup().addFields(elementType).named(list.getFieldName(0));
+    }
+    Type listType = listBuilder.named(list.getName());
+
+    return field == null ? listType : listType.withId(field.id());
+  }
+
+  @Override
+  public Type map(GroupType map, Type keyType, Type valueType) {
+    Preconditions.checkArgument(keyType != null && valueType != null,
+        "Map type must have both key field and value field");
+
+    MappedField field = nameMapping.find(currentPath());
+    Type mapType = Types.buildGroup(map.getRepetition())
+        // Spark 3.1 uses Parquet 1.10 which does not have LogicalTypeAnnotation
+//        .as(LogicalTypeAnnotation.mapType())
```

Comment:
I don't think so.  This should be equivalent but uses deprecated APIs on newer Parquet versions.  If those APIs are eventually removed then it would be an issue that would require shims.